pkg/neorpc/rpcevent/filter.go

```diff
@@ -66,7 +68,29 @@ func Matches(f Comparator, r Container) bool {
 		notification := r.EventPayload().(*state.ContainedNotificationEvent)
 		hashOk := filt.Contract == nil || notification.ScriptHash.Equals(*filt.Contract)
 		nameOk := filt.Name == nil || notification.Name == *filt.Name
-		return hashOk && nameOk
+		parametersOk := true
+		if len(filt.Parameters) > 0 {
+			stackItems := notification.Item.Value().([]stackitem.Item)
+			for i, p := range filt.Parameters {
+				if p.Type == smartcontract.AnyType && p.Value == nil {
+					continue
+				}
+				if i >= len(stackItems) {
+					parametersOk = false
+					break
+				}
+				converted, err := p.ToStackItem()
```

Comment:
It's too expensive to convert every filter's parameter every time you need to access it. Consider contract that emits 100500 notifications that match filter's requirements. To improve it define an additional method `*neorpc.NotificationFilter.ParametersSI() []stackitem.Item`, this method should convert filter parameters to stckitems and cache the resulting value inside filter's structure. Cache should be reused for subsequent invocations of this method. Cache should be cleaned on filter's Copy.

Also, prior to parameter's value comparison use parameter types comparison:
https://github.com/nspcc-dev/neo-go/blob/990634a43af5be0609270b2fce6b103f92686998/pkg/smartcontract/param_type.go#L193

Reply:
Also please add a separate unit-test for various parameter types matching comparison.

Reply:
> isn't stackitem.Item.Equal enough?

It's enough but type comparison allows to fail fast. Parameter to stackitem conversion is not cheap.

> can, please, explain, why it is needed?

Fail fast in case of types mismatch.

Reply:
> Fail fast

got you